.github/workflows/main.yml

```diff
@@ -19,16 +19,21 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      # TODO: cache all the setup
-
       - name: Setup directories
         run: |
           # create directory that will be on the PATH
           mkdir -p ~/.local/bin
           echo "${HOME}/.local/bin" >> $GITHUB_PATH
           mkdir -p ~/source
+      
+      - uses: actions/cache@v2
+        id: cache
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
```

Comment:
This only caches pip dependencies eh? You'll need to use a separate cache step with a separate id for each step you want to skip based on caching. The general rule as I see it is, whatever directories need to be cached go in `path`, optionally line break separated if there's multiple of them:
```
path: |
  ~/foo
  ~/bar
```
Then the key should be created by concatenating the os (`${{ runner.os }}`), the name of the thing you're caching, and then `hashFiles` on any files that determine whether the cache should be invalidated. I.e., for `pip`, any requirements.txt files in the repo could change what dependencies get installed, so we call `hashFiles` on all the requirements.txt files.